core/src/main/java/org/elasticsearch/common/joda/DateMathParser.java

```diff
@@ -190,15 +187,29 @@ private long parseMath(String mathString, long time, boolean roundUp, DateTimeZo
         return dateTime.getMillis();
     }
 
-    private long parseDateTime(String value, DateTimeZone timeZone) {
+    private long parseDateTime(String value, DateTimeZone timeZone, boolean roundUp) {
         DateTimeFormatter parser = dateTimeFormatter.parser();
         if (timeZone != null) {
             parser = parser.withZone(timeZone);
         }
         try {
-            return parser.parseMillis(value);
+            MutableDateTime date;
+            // We use 01/01/1970 as a base date so that things keep working with date
+            // fields that are filled with times without dates
+            if (roundUp) {
+                date = new MutableDateTime(1970, 1, 1, 23, 59, 59, 999, DateTimeZone.UTC);
```

Comment:
This is totally funky and the parsing code is crazy, but I verified it works with weirder timezones.

```
        assertDateMathEquals("2014-11-18", "2014-11-18T00:00:00-09:30", 0, false, DateTimeZone.forID("Pacific/Marquesas"));
        assertDateMathEquals("2014-11-18", "2014-11-18T23:59:59.999-09:30", 0, true, DateTimeZone.forID("Pacific/Marquesas"));
```